File: FantasyRadio/src/main/java/ru/sigil/fantasyradio/saved/SavedFragment.java
package ru.sigil.fantasyradio.saved;

import android.app.AlertDialog;
import android.appwidget.AppWidgetManager;
import android.content.ComponentName;
import android.content.Intent;
import android.os.Bundle;
import android.os.Handler;
import android.os.Message;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageButton;
import android.widget.SeekBar;

import java.io.File;
import java.util.HashMap;
import java.util.Timer;
import java.util.TimerTask;

import javax.inject.Inject;

import ru.sigil.bassplayerlib.IPlayer;
import ru.sigil.bassplayerlib.ITrack;
import ru.sigil.bassplayerlib.PlayState;
import ru.sigil.bassplayerlib.listeners.IEndSyncListener;
import ru.sigil.bassplayerlib.listeners.IPlayStateChangedListener;
import ru.sigil.bassplayerlib.listeners.IVolumeChangedListener;
import ru.sigil.fantasyradio.AbstractListFragment;
import ru.sigil.fantasyradio.R;
import ru.sigil.fantasyradio.dagger.Bootstrap;
import ru.sigil.fantasyradio.utils.RadioStream;
import ru.sigil.fantasyradio.widget.FantasyRadioWidgetProvider;

public class SavedFragment extends AbstractListFragment {
    private MP3Entity mp3EntityForDelete;
    private MP3ArrayAdapter adapter;
    private View savedActivityView;
    @Inject
    MP3Collection mp3Collection;
    @Inject
    IPlayer<RadioStream> player;

    private TimerTask seekTask = new TimerTask() {
        public void run() {
            mp3ProgressHandler.sendEmptyMessage(0);
        }
    };
    private Timer seekTimer = new Timer();

    public SavedFragment() {
        Bootstrap.INSTANCE.getBootstrap().inject(this);
    }

    public void notifyAdapter() {
        adapter = new MP3ArrayAdapter(getActivity().getBaseContext(),
                mp3Collection.getCursor(),
                this::deleteClick, this::playClick
        );
        getLv().setAdapter(adapter);
    }

    private Handler mp3ProgressHandler = new Handler() {
        @Override
        public void handleMessage(Message msg) {
            long streamProgress = player.getProgress();
            try {
                if (player.currentState() == PlayState.PLAY_FILE) {
                    SeekBar sb = getLv().findViewWithTag(player.getCurrentMP3Entity().getDirectory());
                    if (sb != null) {
                        sb.setProgress((int) streamProgress);
                    }
                }
            } catch (NullPointerException e) {
                e.printStackTrace();
            }
        }
    };

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        savedActivityView = inflater.inflate(R.layout.mp3s, container, false);
        player.addPlayStateChangedListener(playStateChangedListener);
        player.addVolumeChangedListener(volumeChangedListener);
        setLv(savedActivityView.findViewById(R.id.MP3ListView));
        return savedActivityView;
    }

    @Override
    public void onDestroyView() {
        player.removePlayStateChangedListener(playStateChangedListener);
        player.removeVolumeChangedListener(volumeChangedListener);
        super.onDestroyView();
    }

    @Override
    public void onResume() {
        notifyAdapter();
        super.onResume();
    }


    public void playClick(View v) {
        updateWidget();
        if (player.isPaused()) {
            if (player.getCurrentMP3Entity() != null && v.getTag() != null
                    && player.getCurrentMP3Entity().getDirectory().equals(((MP3Entity) v.getTag()).getDirectory())) {
                // Это была пауза.
                player.resume();
                return;
            }
        }
        try {
            seekTimer.scheduleAtFixedRate(seekTask, 0, 1000);
        } catch (Exception e) {
            e.printStackTrace();
        }
        if (player.currentState() == PlayState.PLAY || player.currentState() == PlayState.PLAY_FILE) {
            if (player.getCurrentMP3Entity() != null && v.getTag() != null
                    && player.getCurrentMP3Entity().getDirectory().equals(((MP3Entity) v.getTag()).getDirectory())) {
                //TODO !!!
                player.pause();
                ImageButton bv = (ImageButton) v;// !!!!!!!!!!!!!!!
                bv.setImageResource(R.drawable.play_states);
            } else {
                // Нажата кнопка плэй у другого трека
                player.playFile((MP3Entity) v.getTag());
                adapter.notifyDataSetChanged();
            }
        } else {
            player.playFile((MP3Entity) v.getTag());
            // -------------------------------------------------
            adapter.notifyDataSetChanged();
        }
    }

    public void deleteClick(View v) {// Тут удаляем mp3
        HashMap<String, String> messageMap;
        messageMap = (HashMap<String, String>) v.getTag();
        mp3EntityForDelete = new MP3Entity(messageMap.get("artist"), messageMap.get("title"),
                messageMap.get("directory"), messageMap.get("time"));
        AlertDialog.Builder alertDialogBuilder = new AlertDialog.Builder(getActivity());
        alertDialogBuilder
                .setTitle(getString(R.string.are_you_sure_want_delete));
        alertDialogBuilder.setPositiveButton(getString(R.string.yes),
                (dialog, which) -> {
                    mp3Collection.remove(
                            mp3EntityForDelete);
                    if (player.getCurrentMP3Entity() != null) {
                        if (mp3EntityForDelete.getDirectory().equals(player.getCurrentMP3Entity().getDirectory())) {
                            player.stop();
                        }
                    }
                    File f = new File(mp3EntityForDelete.getDirectory());
                    Log.v("dir", mp3EntityForDelete.getDirectory());
                    Boolean b = f.delete();
                    Log.v("tf", b.toString());
                    dialog.dismiss();
                    onResume();
                });
        alertDialogBuilder.setNegativeButton(getString(R.string.no),
                (dialog, which) -> {
                    // here you can add functions
                    dialog.dismiss();
                });
        alertDialogBuilder.create().show();
        this.onResume();
    }

    private void updateWidget() {
        Intent intent = new Intent(getActivity().getApplicationContext(), FantasyRadioWidgetProvider.class);
        intent.setAction(AppWidgetManager.ACTION_APPWIDGET_UPDATE);
        int ids[] = AppWidgetManager.getInstance(getActivity().getApplicationContext()).getAppWidgetIds(
                new ComponentName(getActivity().getApplicationContext(), FantasyRadioWidgetProvider.class));
        intent.putExtra(AppWidgetManager.EXTRA_APPWIDGET_IDS, ids);
        getActivity().getApplicationContext().sendBroadcast(intent);
    }

    private final IPlayStateChangedListener playStateChangedListener = (playState) -> adapter.notifyDataSetChanged();


    private IVolumeChangedListener volumeChangedListener = (volume) -> getActivity().runOnUiThread(() -> {
        if (player.getCurrentMP3Entity() != null) {
            SeekBar volumeSeekBar = getLv().findViewWithTag(player.getCurrentMP3Entity().getDirectory() + "volume");
            volumeSeekBar.setProgress((int) (volume * 100));
        }
    });
}

File: FantasyRadio/src/main/kotlin/ru/sigil/fantasyradio/saved/SavedFragment.kt
package ru.sigil.fantasyradio.saved

import android.app.AlertDialog
import android.os.Handler
import android.os.Message
import android.view.View
import ru.sigil.fantasyradio.AbstractListFragment
import ru.sigil.fantasyradio.dagger.Bootstrap
import ru.sigil.fantasyradio.utils.RadioStream
import ru.sigil.bassplayerlib.IPlayer
import javax.inject.Inject
import android.widget.SeekBar
import ru.sigil.bassplayerlib.PlayState
import android.os.Bundle
import android.view.ViewGroup
import android.view.LayoutInflater
import ru.sigil.fantasyradio.R
import android.widget.ImageButton
import android.appwidget.AppWidgetManager
import ru.sigil.fantasyradio.widget.FantasyRadioWidgetProvider
import android.content.ComponentName
import android.content.Intent
import android.util.Log
import ru.sigil.bassplayerlib.listeners.IPlayStateChangedListener
import ru.sigil.bassplayerlib.listeners.IVolumeChangedListener
import java.io.File
import java.util.TimerTask
import java.util.Timer
import kotlin.collections.HashMap

/**
 * Created by namelessone
 * on 02.12.18.
 */
class SavedFragment: AbstractListFragment() {
    init {
        Bootstrap.INSTANCE.getBootstrap().inject(this)
    }

    private var mp3EntityForDelete: MP3Entity? = null
    private var adapter: MP3ArrayAdapter? = null
    private var savedActivityView: View? = null
    @Inject
    lateinit var mp3Collection: MP3Collection
    @Inject
    lateinit var player: IPlayer<RadioStream>

    private val seekTask = object : TimerTask() {
        override fun run() {
            mp3ProgressHandler.sendEmptyMessage(0)
        }
    }

    private val seekTimer = Timer()

    fun notifyAdapter() {
        adapter = MP3ArrayAdapter(activity!!.baseContext,
                mp3Collection.getCursor()!!,
                View.OnClickListener(this::deleteClick), View.OnClickListener(this::playClick)
        )
        lv?.adapter = adapter
    }

    private val mp3ProgressHandler = object : Handler() {
        override fun handleMessage(msg: Message) {
            val streamProgress = player?.progress
            try {
                if (player?.playState === PlayState.PLAY_FILE) {
                    val sb = lv?.findViewWithTag<SeekBar>(player?.currentMP3Entity?.directory)
                    sb?.progress = streamProgress!!.toInt()
                }
            } catch (e: NullPointerException) {
                e.printStackTrace()
            }

        }
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        savedActivityView = inflater.inflate(R.layout.mp3s, container, false)
        player.addPlayStateChangedListener(playStateChangedListener)
        player.addVolumeChangedListener(volumeChangedListener)
        lv = savedActivityView?.findViewById(R.id.MP3ListView)
        return savedActivityView
    }

    override fun onDestroyView() {
        player.removePlayStateChangedListener(playStateChangedListener)
        player.removeVolumeChangedListener(volumeChangedListener)
        super.onDestroyView()
    }

    override fun onResume() {
        notifyAdapter()
        super.onResume()
    }

    fun playClick(v: View) {
        updateWidget()
        if (player.isPaused) {
            if (player.currentMP3Entity != null && v.tag != null
                    && player.currentMP3Entity?.directory.equals((v.tag as MP3Entity).directory)) {
                // Это была пауза.
                player.resume()
                return
            }
        }
        try {
            seekTimer.scheduleAtFixedRate(seekTask, 0, 1000)
        } catch (e: Exception) {
            e.printStackTrace()
        }

        if (player.playState === PlayState.PLAY || player.playState === PlayState.PLAY_FILE) {
            if (player.currentMP3Entity != null && v.tag != null
                    && player.currentMP3Entity?.directory == (v.tag as MP3Entity).directory) {
                //TODO !!!
                player.pause()
                val bv = v as ImageButton// !!!!!!!!!!!!!!!
                bv.setImageResource(R.drawable.play_states)
            } else {
                // Нажата кнопка плэй у другого трека
                player.playFile(v.tag as MP3Entity)
                adapter?.notifyDataSetChanged()
            }
        } else {
            player.playFile(v.tag as MP3Entity)
            adapter?.notifyDataSetChanged()
        }
    }

    fun deleteClick(v: View) {// Тут удаляем mp3
        val messageMap: HashMap<String, String> = v.tag as HashMap<String, String>
        mp3EntityForDelete = MP3Entity(messageMap["artist"], messageMap["title"],
                messageMap["directory"]!!, messageMap["time"])
        val alertDialogBuilder = AlertDialog.Builder(activity)
        alertDialogBuilder
                .setTitle(getString(R.string.are_you_sure_want_delete))
        alertDialogBuilder.setPositiveButton(getString(R.string.yes)
        ) { dialog, _ ->
            mp3Collection.remove(mp3EntityForDelete ?: return@setPositiveButton) //TODO хрень какая-то. Разобраться
            if (player.currentMP3Entity != null) {
                if (mp3EntityForDelete?.directory.equals(player.currentMP3Entity?.directory)) {
                    player.stop()
                }
            }
            val f = File(mp3EntityForDelete?.directory)
            Log.v("dir", mp3EntityForDelete?.directory ?: "")
            val b = f.delete()
            Log.v("tf", b.toString())
            dialog.dismiss()
            onResume()
        }
        alertDialogBuilder.setNegativeButton(getString(R.string.no)
        ) { dialog, _ ->
            // here you can add functions
            dialog.dismiss()
        }
        alertDialogBuilder.create().show()
        this.onResume()
    }

    private fun updateWidget() {
        val intent = Intent(activity!!.applicationContext, FantasyRadioWidgetProvider::class.java)
        intent.action = AppWidgetManager.ACTION_APPWIDGET_UPDATE
        val ids = AppWidgetManager.getInstance(activity!!.applicationContext).getAppWidgetIds(
                ComponentName(activity!!.applicationContext, FantasyRadioWidgetProvider::class.java))
        intent.putExtra(AppWidgetManager.EXTRA_APPWIDGET_IDS, ids)
        activity?.applicationContext?.sendBroadcast(intent)
    }

    private val playStateChangedListener = object : IPlayStateChangedListener {
        override fun onPlayStateChanged(playState: PlayState) {
            activity?.runOnUiThread { adapter?.notifyDataSetChanged() }
         }
    }

    private val volumeChangedListener = object : IVolumeChangedListener {
        override fun onVolumeChanged(volume: Float) {
            activity?.runOnUiThread {
                if (player.currentMP3Entity != null) {
                    val volumeSeekBar = lv?.findViewWithTag<SeekBar>(player.currentMP3Entity?.directory + "volume")
                    volumeSeekBar?.progress = (volume * 100).toInt()
                }
            }
        }
    }
}